docs/_rpc/ns-debug.md

```diff
@@ -744,11 +744,15 @@ specifies the options for this specific call. The possible options are:
 
 #### JavaScript-based tracing
 
-Specifying the `tracer` option in the second argument enables JavaScript-based tracing. In this mode, `tracer` is interpreted as a JavaScript expression that is expected to evaluate to an object which must expose the `result` and `fault` methods. There exist 3 additional methods, namely: `step`, `enter` and `exit`. You must provide either `step`, or `enter` AND `exit` (i.e. these two must be exposed together). You may expose all three if you choose to do so.
+Specifying the `tracer` option in the second argument enables JavaScript-based tracing. In this mode, `tracer` is interpreted as a JavaScript expression that is expected to evaluate to an object which must expose the `result` and `fault` methods. There exist 4 additional methods, namely: `setup`, `step`, `enter`, and `exit`. `enter` and `exit` must be present or omitted together.
+
+##### Setup
+
+`setup` is invoked once, in the beginning when the tracer is being constructed by Geth for a given transaction. It takes in one argument `config`. `config` allows users to pass in options to the tracer. `config` is to be JSON-decoded for usage and its default value is `"{}"`.
```

Comment:
@s1na Can you please explain what these new config options are?

Reply:
These configs are meant to be tracer-specific. I.e. each JS or native tracer can now define which options it accepts. Let's say the `callTracer` has the following option to only collect data about the top-level call:

```
debug.traceTransaction('<txhash>, { tracer: 'callTracer', tracerConfig: { onlyTopCall: true } })
```

Or we're contemplating adding the following option to the `prestateTracer` to also collect the state of each account *after* tx execution:

```
debug.traceTransaction('<txhash>, { tracer: 'prestateTracer': tracerConfig: { collectPostState: true } })
```

But passing `collectPostState` to `callTracer` will have no effect because `callTracer` doesn't parse/understand it.